modules/xfem/src/efa/EFAElement.C

```diff
@@ -0,0 +1,366 @@
+                        <<" from PermanentNodes, but couldn't find it");
+          }
+          childOfNeighborNode = childNode;
+        }
+        else
+        {
+          EFAError("Attempting to merge nodes: "<<childNode->id()<<" and "
+                      <<childOfNeighborNode->id()<<" but both are permanent themselves");
+        }
+      }
+      else
+      {
+        if (childOfNeighborNode->parent() != childNode &&
+            childOfNeighborNode->parent() != childNode->parent())
+        {
+          EFAError("Attempting to merge nodes "<<childOfNeighborNode->idCatString()<<" and "
+                      <<childNode->idCatString()<<" but neither the 2nd node nor its parent is parent of the 1st");
+        }
+        childOfNeighborElem->switchNode(childNode, childOfNeighborNode, true);
+        if (!deleteFromMap(TempNodes, childOfNeighborNode))
+        {
+          EFAError("Attempted to delete node: "<<childOfNeighborNode->id()<<" from TempNodes, but couldn't find it");
+        }
+        childOfNeighborNode = childNode;
+      }
+    }
+    else if (childOfNeighborNode->category() == N_CATEGORY_PERMANENT)
+    {
+      if (childNode->parent() != childOfNeighborNode &&
+          childNode->parent() != childOfNeighborNode->parent())
+      {
+        EFAError("Attempting to merge nodes "<<childNode->id()<<" and "
+                    <<childOfNeighborNode->id()<<" but neither the 2nd node nor its parent is parent of the 1st");
+      }
+      childElem->switchNode(childOfNeighborNode, childNode, true);
+      if (!deleteFromMap(TempNodes, childNode))
+      {
+        EFAError("Attempted to delete node: "<<childNode->id()<<" from TempNodes, but couldn't find it");
+      }
+      childNode = childOfNeighborNode;
+    }
+    else //both nodes are temporary -- create new permanent node and delete temporary nodes
+    {
+      unsigned int new_node_id = getNewID(PermanentNodes);
+      EFANode* newNode = new EFANode(new_node_id,N_CATEGORY_PERMANENT,childNode->parent());
+      PermanentNodes.insert(std::make_pair(new_node_id,newNode));
+
+      childOfNeighborElem->switchNode(newNode, childOfNeighborNode, true);
+      childElem->switchNode(newNode, childNode, true);
+
+      if (childNode->parent() != childOfNeighborNode->parent())
+      {
+        EFAError("Attempting to merge nodes "<<childNode->id()<<" and "
+                    <<childOfNeighborNode->id()<<" but they don't share a common parent");
+      }
+
+      if (!deleteFromMap(TempNodes, childOfNeighborNode))
+      {
+        EFAError("Attempted to delete node: "<<childOfNeighborNode->id()<<" from TempNodes, but couldn't find it");
+      }
+      if (!deleteFromMap(TempNodes, childNode))
+      {
+        EFAError("Attempted to delete node: "<<childNode->id()<<" from TempNodes, but couldn't find it");
+      }
+      childOfNeighborNode = newNode;
+      childNode = newNode;
+    }
+  }
+}
```

Comment:
Hmm.  I'll have to ponder how I could split it into two PRs and still have it work.  Maybe I can do that.